### DISCUSSION TITLE
NOTHING WORKS

### DISCUSSION THREAD

**Discussion by NorthBaandee** [timmbogner/Farm-Data-Relay-System#185] [Q&A]
The examples as published upload successfully to my esp32s (I have TTGO_OLED_LORA and NodeMCU-32S boards) but nothing is ever displayed on the OLEDs or received in the serial monitor. The NodeMCU-32S just gets hot. There is nothing wrong with the boards. I can run most other sketches without problems.

**Comment by timmbogner** [comment]
Hi @NorthBaandee ,
I've compiled and uploaded the 1_UART_Gateway sketch to a variety of ESP32 boards, including the TTGO LoRa32 V1 and they all send some form of the following message over the serial port after boot:
```
19:03:33.318 -> ets Jun  8 2016 00:22:57
19:03:33.318 -> 
19:03:33.318 -> rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
19:03:33.318 -> configsip: 0, SPIWP:0xee
19:03:33.318 -> clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
19:03:33.318 -> mode:DIO, clock div:1
19:03:33.318 -> load:0x3fff0030,len:1344
19:03:33.358 -> load:0x40078000,len:13964
19:03:33.358 -> load:0x40080400,len:3600
19:03:33.358 -> entry 0x400805f0
19:03:33.612 -> Address:1
19:03:33.612 -> Initializing ESP-NOW!
19:03:33.679 ->  ESP-NOW Initialized.
```

I am using the newest version of both Arduino IDE and the ESP32 board definitions.  Without more details, I can't really tell you why it's not working on your end.  What exactly does the serial monitor display when your ESP32 is booting?

Note that the OLED screen is not enabled by default, and should be configured and enabled using these configs:
```
// OLED -- Displays console debugging messages on an SSD1306 I²C OLED
#define USE_OLED    
#define OLED_HEADER "FDRS"
#define OLED_PAGE_SECS 30
#define OLED_SDA 4
#define OLED_SCL 15
#define OLED_RST 16
```

**Reply by NorthBaandee** [reply]
Hi Tim thanks for getting back to me so promptly
Here is the output of compiling and uploading:

Sketch uses 747817 bytes (57%) of program storage space. Maximum is 1310720 bytes.
Global variables use 45624 bytes (13%) of dynamic memory, leaving 282056 bytes for local variables. Maximum is 327680 bytes.
esptool.py v4.2.1
Serial port /dev/ttyUSB0
Connecting.......
Chip is ESP32-D0WD-V3 (revision 3)
Features: WiFi, BT, Dual Core, 240MHz, VRef calibration in efuse, Coding Scheme None
Crystal is 40MHz
MAC: e0:5a:1b:6c:ae:e8
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Erasing flash (this may take a while)...
Chip erase completed successfully in 3.1s
Compressed 17440 bytes to 12108...
Writing at 0x00001000... (100 %)
Wrote 17440 bytes (12108 compressed) at 0x00001000 in 1.2 seconds (effective 114.4 kbit/s)...
Hash of data verified.
Compressed 3072 bytes to 146...
Writing at 0x00008000... (100 %)
Wrote 3072 bytes (146 compressed) at 0x00008000 in 0.0 seconds (effective 565.6 kbit/s)...
Hash of data verified.
Compressed 8192 bytes to 47...
Writing at 0x0000e000... (100 %)
Wrote 8192 bytes (47 compressed) at 0x0000e000 in 0.1 seconds (effective 1008.0 kbit/s)...
Hash of data verified.
Compressed 753584 bytes to 480508...
Writing at 0x00010000... (3 %)
Writing at 0x0001b855... (6 %)
Writing at 0x0002902c... (10 %)
Writing at 0x0003331b... (13 %)
Writing at 0x0003a05f... (16 %)
Writing at 0x0003f801... (20 %)
Writing at 0x00044d00... (23 %)
Writing at 0x0004a0d4... (26 %)
Writing at 0x0004f320... (30 %)
Writing at 0x000544df... (33 %)
Writing at 0x000597a4... (36 %)
Writing at 0x0005eaad... (40 %)
Writing at 0x00063f51... (43 %)
Writing at 0x0006915d... (46 %)
Writing at 0x0006ed61... (50 %)
Writing at 0x00074681... (53 %)
Writing at 0x00079ab9... (56 %)
Writing at 0x0007eceb... (60 %)
Writing at 0x000840db... (63 %)
Writing at 0x00089620... (66 %)
Writing at 0x0008ecd7... (70 %)
Writing at 0x00094483... (73 %)
Writing at 0x0009a211... (76 %)
Writing at 0x0009fbfb... (80 %)
Writing at 0x000a80a6... (83 %)
Writing at 0x000b04f0... (86 %)
Writing at 0x000b57b3... (90 %)
Writing at 0x000bb1b5... (93 %)
Writing at 0x000c087d... (96 %)
Writing at 0x000c60d3... (100 %)
Wrote 753584 bytes (480508 compressed) at 0x00010000 in 42.2 seconds (effective 142.9 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...

The serial monitor indicates that the board is connected but no output whatever appears. The red LED on the board (the one closest the edge) shines steady and bright.

regards PeterB

**Reply by timmbogner** [reply]
Any time the ESP32 boots, it sends some data out over it's serial port. If you aren't able see some data on the serial monitor every time you reset the ESP32 module, that indicates you have an issue between your module and your computer.   For the sake of testing, try running this  example sketch from Espressif and see if anything changes: ```Examples/ESP32/ResetReason```

**Reply by NorthBaandee** [reply]
serial output of reset reason:
10:40:31.457 -> CPU0 reset reason:
10:40:31.457 -> POWERON_RESET
10:40:31.457 -> Vbat power on reset
10:40:31.457 -> CPU1 reset reason:
10:40:31.457 -> EXT_CPU_RESET
10:40:31.457 -> for APP CPU, reseted by PRO CPU
10:40:31.457 -> Going to sleep

**Reply by timmbogner** [reply]
>There is something wacky going on with serial now, that never previously occured. I have a sketch that sets up an access point which controls four relays. The access point starts, I can join it, load it's web page and control the relays, but I see nothing on the serial monitor. This is new behaviour. Previously it showed the IP, the relay states etc. Now I see absolutely nothing. I just uploaded a sketch which does nothing more than Serial.println("hello"); over and over. nothing.

It sounds like intermittent serial interruptions, which could be as simple as replacing your USB cable.

**Reply by timmbogner** [reply]
Also could be something with USB permissions in Linux. I've dealt with that in Ubuntu but it's been a while.

**Reply by NorthBaandee** [reply]
Yep I don't get it at all. I'm on linux Mint. BTW just to add to my woes I cannot get grafana to work on raspi zeroW. The whole package seems to be seriously broken, and either segfaults or "performs illegal instructions". Maybe Trump has been tampering with it.
anyway I think you are antipodal to me and therefore 12hours time difference.

**Comment by svdrummer** [comment]
Is this one of those boards with USB built in like the S3 boards,? I had issues with serial. it is a config setting on the IDE

**Reply by NorthBaandee** [reply]
What sort of issues? This board works perfectly with every other sketch i have uploaded to it. Ive had it controlling 4 mains power outlet relays and correctly reporting it over /dev/ttyUSB0.
I have noticed that arduino 2 sometimes cannot decide whether usb is on  /dev/ttyAMA0 or /dev/ttyUSB0 but that is not a big deal.

**Reply by svdrummer** [reply]
Just trying to rule out an issue. following the fault. Ie, the code works for Tim, not for you. Hence the issue could be at your end. Some chips dont release the USB programming to the serial port, unless it is set. As you have your board working with other projects, then that must be okay.

**Reply by NorthBaandee** [reply]
There is something wacky going on with serial now, that never previously occured. I have a sketch that sets up an access point which controls four relays. The access point starts, I can join it, load it's web page and control the relays, but I see nothing on the serial monitor. This is new behaviour. Previously it showed the IP, the relay states etc. Now I see absolutely nothing. I just uploaded a sketch which does nothing more than Serial.println("hello"); over and over. nothing.

**Comment by svdrummer** [comment]
Check 

USB CDC On Boot | Enabled

In the config section on the IDE. This tells the programming port to switch to serial monitor after programming. If disables, the serial monitor is blank.

**Reply by timmbogner** [reply]
This is a really great tip for the S and C-series chips! Note that you won't see it when a 'regular' ESP32 board (for example: "ESP32 Dev Module") is selected.

I think this just saved me from some frustration with my LilyGo T3S3 boards, though. Much appreciated!

**Reply by NorthBaandee** [reply]
I have no idea how to change that setting in Arduino-IDE since I am quite unable to find a config section. I can see a "preferences" menu but nothing there helps me.
Anyway maybe just looking at the words in your replay is sufficient to make it work because I now have this output on the serial monitor:

09:46:24.828 -> ed
09:46:24.828 -> 
09:46:24.828 -> [   175][D][WiFiGeneric.cpp:rduino Event: 3 - STA_ST$z}J��Incoming Serial.
09:46:25.783 -> Sending to ESP-NOW Neighbor #2
09:46:25.814 -> Sending to ESP-NOW peers.

Thanks again for your support and amazing telepathic abilities from the other side of the planet!
warm regards
PeterB

**Reply by NorthBaandee** [reply]
....so I set up a BMP280 sensor which relays to a lora repeater to the serial gateway. This is what it is reporting:

10:43:19.326 -> Incoming ESP-NOW.
10:43:19.326 -> Sending Serial.
10:43:19.326 -> [{"id":1,"type":1,"data":182.8699951},{"id":1,"type":4,"data":577.4482422}]

The temp and pressure readings are a bit warm, and the barometer appears to be on the moon?. When I read the serial data directly from the sensor its 29 degrees and 997hPa.

**Reply by shaffenmeister** [reply]
Hi,

share some code please. This is most probably an issue with your code and not the FDRS.

When Choosing e.g. an ESP32S3 board, the Arduino config dialog should look like this:

![image](https://github.com/timmbogner/Farm-Data-Relay-System/assets/10982665/fff8b9ca-2844-4497-84d5-7256003583da)

Make sure to choose the right ports (Arduino on Linux):
![image](https://github.com/timmbogner/Farm-Data-Relay-System/assets/10982665/4df802d2-f737-4ce3-af0e-a5aedb72665a)

Check USB settings for your board:
![image](https://github.com/timmbogner/Farm-Data-Relay-System/assets/10982665/4259c98d-c458-4cf0-8e75-5deea9c4037a)

This may seem overly complicated but should point you in the right direction. There are numerous guides on the internet which explain on how to use ESP32S2, ESPS3 and the like with Arduino.

HTH

**Reply by timmbogner** [reply]
What exactly does the serial output of your sensor look like?

I've hooked up a BMP280 to an ESP8266 and mine looks like this:
```
01:43:06.614 ->  Reading ID 1
01:43:06.614 ->  Gateway: 1
01:43:06.614 -> Powering up the sensor array!
01:43:06.655 -> Initializing ESP-NOW!
01:43:06.777 ->  ESP-NOW Initialized.
01:43:06.850 -> Id: 1 - Type: 1 - Data loaded: 23.31
01:43:06.850 -> Id: 1 - Type: 4 - Data loaded: 993.20
01:43:06.893 -> Sending FDRS Packet!
01:43:06.893 ->  Deep sleeping.
```

**Reply by NorthBaandee** [reply]
Sorry guys the problem was the GND pin to the sensor had come loose.  I'm almost fully up and running now but still not sure what the issue is with /dev/ttyUSB0 on my desktop pc. Doing all my work on the wife's laptop now seems to be less frustrating. Thanks for all your help. I learned a lot. Just hoping the boffins over at grafana can sort out their raspberry pi binaries some time soon and then the picture is complete 😎

**Reply by timmbogner** [reply]
Glad you got it working. Something to investigate is whether some other program might be using the COM ports. For example: any time I have Ultimaker Cura open in the background, the Arduino serial monitor won't be able to connect to devices.

From what I've gathered, the Pi Zero just isn't powerful enough to run Grafana, especially if you're running a database as well. It can barely handle a Node-RED dashboard with a few charts 🙄
You may have better luck with the Zero 2, but you'd probably save yourself a lot of frustration if you bumped up to bigger Pi or an old laptop/workstation for a server.

**Reply by NorthBaandee** [reply]
I had a look at top on the zeroW and yeah influxdb really hogs it. So i have disable it for the tome being.
I have configured it so that /dev/ttyAMA0 is on pins 8 and 10 and if i wire them to the TX and RX pins on the nodeMCU32-S i can see the output of ResetReason.ino by doing cat /dev/ttyAMA0
However if I load the frds uart gateway sketch nothing is received. 
Looking at the fdrs_gateway_serial.h file there are no RXD2 and TXD2 pin definitions matching the nodeMCU32-S, which is TX on GPIO1 and RX on GPIO3. 
Or am i barking up the wrong tree here?

**Reply by timmbogner** [reply]
I'm not sure what's up with your serial ports. I use a Win10 machine for the most part, so I'm not sure where to start debugging that. I'll revisit this if I get reports of more UART issues with FDRS.

In case it helps here is the overview of UART (serial) works on the ESP32 gateway:

The ESP32 chip has three UART interfaces, two of which may be used in FDRS. The pin assignments of all three can be changed, but the first usually shouldn't be.
- The first, '**Serial**' or _UART0_ is the serial port used for the USB interface. It is the interface that generally programs the chip, as well as displays messages in the serial monitor.  It is by default connected to GPIO3 and GPIO1, and in most cases it is physically connected to USB-UART chip via these two pins. If you change this assignment, your serial monitor _will_ go blank. This is because the pins that the USB interface is connected to are no longer being used for UART. In FDRS, this interface will display debugging information as long as FDRS_DEBUG is defined. It will *always* display any JSON data being sent by the `sendSerial()` function.

- The second interface,  '**Serial1**' or _UART1_ is the one which can be configured within FDRS. This interface does not display debugging information, and only prints JSON data. Define them with `TXD2` and `RXD2` in the gateway config file, or FDRS will choose default pins that are friendly to your chip.

**Reply by NorthBaandee** [reply]
The serial ports are driving me insane. I got rxd2/txd2 working on the nodeMCU32-s by setting them to RXD2 = gpio16 and TXD2 = gpio17. I cannot understand why some sketches produce reliable transmissions on USB serial and others do not. Perhaps there is an issue with nodeMCU32-s boards. The ttgo boards don't seem to have any problems. Another thing i have noticed is the lora range seems poor using radiolib at 866mhz 17db compared to the sandeep library. (Yes i know we are supposed to use 915 in australia but i only have 866 antennae at the moment). I might do some more testing once i get my basic setup more reliable. Just need to modify the lora gateway sketch to send rx dbi telemetry i think. Thanks for putting up with me. I'm a bit of a noob

**Reply by svdrummer** [reply]
Just as a side, if you are closer to cities, Telstra monitor frequencies as they have the rights to those. I think they are setting up a public/for a fee LORA/IOT network. Fines from ACCC have been threatened. I am waiting on the correct frequency lora at the moment. I am not happy about Telstra sticking it's nose in.

**Reply by NorthBaandee** [reply]
I am in a very remote location so not worried about telstra yet. 433 should give the best range. In my experience as a radio amateur ACMA will only move if there is a complaint, and even then they will try for a technical rather than a punitive solution.
